content/posts/wasi-experimental-http.md

```diff
@@ -0,0 +1,146 @@
+---
+title:
+  "An experimental outbound HTTP library for the WebAssembly System Interface"
+description:
+  "Send HTTP requests from Rust and AssemblyScript Wasm modules running in
+  Wasmtime"
+date: 2021-03-08 00:00:00 +0000 UTC
+authorname: "Radu Matei"
+author: "@matei_radu"
+authorlink: "https://twitter.com/matei_radu"
+image: "images/twitter-card.png"
+tags: ["wasm", "wasi"]
+---
+
+Over the last year, our team has been experimenting with executing WebAssembly
+workloads on the server using the [WebAssembly System Interface][wasi] and
+[Wasmtime][wasmtime], and while more and more [proposals][proposals] get
+implemented, networking is one scenario that doesn't have a stable API yet,
+which limits current applications compiled to WASI, and restricts the types of
+workloads that can be executed in WASI runtimes.
+
+Today, [we are releasing a library][gh] which adds experimental and _temporary_
+support for outgoing HTTP requests for WASI modules running in Wasmtime. It is
+important to note from the beginning that this library does not aim to replace
+the WASI sockets API, but rather provide a temporary workaround until the WASI
+networking API is stable, and we expect that once [the WASI sockets
+proposal][sockets-wip] gets adopted and implemented in language toolchains, the
+need for this library would go away.
+
+### Using the new HTTP library
+
+There are two aspects to a WebAssembly library - building guest Wasm modules,
+and adding runtime support for instantiating and executing those modules. Let's
+first see how to build a guest module in Rust:
+
+```rust
+use bytes::Bytes;
+use http;
+use wasi_experimental_http;
+
+#[no_mangle]
+pub extern "C" fn _start() {
+    let url = "https://<some-domain>/post".to_string();
+    let req = http::request::Builder::new()
+        .method(http::Method::POST)
+        .uri(&url)
+        .header("Content-Type", "text/plain")
+    let b = Bytes::from("sending a body in a POST request");
+    let req = req.body(Some(b)).unwrap();
+
+    let res = wasi_experimental_http::request(req).expect("cannot make request");
+    let str = std::str::from_utf8(&res.body()).unwrap().to_string();
+    // do something with the body
+}
+```
+
+From the start, the one thing to note is that we are reusing the Rust `Request`
```

Comment:
suggest putting your point first: "Note that because we reuse <yadda yadda>, this way of building HTTP requests should feel somewhat familiar to Rust developers and it means that a response can be sent and received [oddly the wrong way arougn? request?] using the familiar `wasi_experimental_http::request(req)`." Direct, to the point. 

Then new graph:
"Now this simple program can be compiled to .... compiles to WebAssembly. The library we are releasing also...."